### PR TITLE
Add `BitAnd` to `mc-sgx-core-types::ReportData`

### PR DESCRIPTION
**Pull request by nick-mobilecoin**

### CODE REVIEW

**Comment by nick-mobilecoin** [comment]
Current dependencies on/for this PR:
* main
  * **PR #291** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx/291" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx/291?utm_source=stack-comment).

**Review by jcape — APPROVED** [review]

File: core/types/src/report.rs [hunk 1]
```diff
@@ -33,6 +34,22 @@ impl_newtype_for_bytestruct! {
...
+/// There are times when only part of [`ReportData`] is of interest. [`BitAnd`]
+/// allows clients to mask off the parts of [`ReportData`] that are not of
+/// interest.
+impl BitAnd for ReportData {
+    type Output = Self;
+
+    fn bitand(self, rhs: Self) -> Self::Output {
+        // NB: Due to use in verification, this must be constant time.
+        let mut output = Self::default();
+        for (i, (a, b)) in self.0.d.iter().zip(rhs.0.d.iter()).enumerate() {
```
Comment (awygle) [inline core/types/src/report.rs:46]: Wow, this is _dense_. I think I understand what's happening here but I think a simple `for i in range` type loop would be a lot more readable.

**Review by awygle — APPROVED** [review]
There's nothing wrong with this but I think the iterator magic is a bit opaque.

File: core/types/src/report.rs (hunk 1 above)
Comment (nick-mobilecoin) [reply core/types/src/report.rs:46]: Do you have a suggested implementation that you're thinking of?
Just for sanity I did this https://godbolt.org/z/W8ae5rPM8 to provide a bit more confidence the compiler won't care either way (if one were micro-optimizing), but also to show constant time ass.

I don't have a preference one way or the other. (though I have some scars from embedded compilers that would re-compute the array offset even in release :()
```rust
        for (i, o) in output.0.d.iter_mut().enumerate() {
            *o = self.0.d[i] & rhs.0.d[i];

        }
```

**Review by samdealy — APPROVED** [review]

File: core/types/src/report.rs (hunk 1 above)
Comment (nick-mobilecoin) [reply core/types/src/report.rs:46]: Going to merge for now and we can discuss how dense we want to get in iterators moving forward

**Pull request merged**